### DISCUSSION TITLE
Using Theseus to generate a full 3D trajectory of a 7DOF robot arm via motion planning

### DISCUSSION THREAD

**Discussion by DanielTakeshi** [facebookresearch/theseus#564] [General]
After going through some of the Theseus tutorials, here's what I can piece together:

- From the tutorials 04 and 05, those use motion planning in a 2D environment to avoid collisions. They produce a trajectory which specifies 2D positions and 2D velocities of the "robot" (which is really a point mass) at each time step. Thus, if a trajectory lasts 100 time steps, Theseus needs to find values for optimization variables of positions and velocities for each of the 100 time steps. This helps it avoid collisions, etc.
- There is an inverse kinematics example: https://github.com/facebookresearch/theseus/blob/main/examples/inverse_kinematics.py  Given a target pose of the robot EE (position and quaternion) we want to find the joint angles that will get the EE to that pose. This produces one set of joint angles for each target, so it's not specifying a full trajectory, it's just producing the IK solution.

Now consider the mix of the two. Instead of a 2D map and a "point mass" robot as in tutorials 04, and 05, what if we have a full robot arm like the Franka, as in the inverse kinematics examples. However, instead of finding one joint angle, we want to find joint angles for all time steps in a trajectory, such that the full robot arm can avoid any collisions we insert in the environment. Here the environment would be 3D with obstacles, so the point of motion planning is to find a full 3D collision-free trajectory.

Questions:

- Would Theseus be a good use case for this idea? I see that GPMP (tutorial 04) tested on a 7DOF arm planning problem https://homes.cs.washington.edu/~bboots/files/GPMP.pdf but this was before Theseus was developed.
- I am assuming there is not a quick example script ready to show for this. Would it make sense to try and build upon the previously mentioned points (the tutorials and the IK example) to implement this idea?

Thanks again for the great library!

**Comment by mhmukadam** [comment]
Hi @DanielTakeshi, that's a great insight, what you suggest is indeed possible.

Q1: Yes, Theseus is a good use case for this, especially if you want a differentiable/batchable planner. The closest prior work you can reference is [GPMP2](https://arxiv.org/abs/1707.07383) (follow up of the GPMP work you mentioned) with its [associated cpp code](https://github.com/borglab/gpmp2) ([matlab tutorial](https://github.com/borglab/gpmp2/blob/master/doc/ExampleMatlab3D.md) and [python example](https://github.com/borglab/gpmp2/blob/master/python/examples/WAMFactorGraphExample.py)).

Q2: This has been something on our stack of things but we haven't been able to prioritize. If you are interested in giving it a try we can discuss some of the details. You can already easily put together a free-space arm planner using the [GPMotionModel](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/motionmodel/double_integrator.py#L179) cost if the goal is in joint space or use FK if the goal is in end effector space. For 3D collision avoidance to build the full arm planner, two key things need to be added: (i) FK with [body points / spheres](https://github.com/borglab/gpmp2/blob/master/gpmp2/kinematics/RobotModel.h#L20) and (ii) extending the [2D collision](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/collision/collision.py#L17) cost and [SDF support](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/collision/signed_distance_field.py#L16) to 3D.

**Reply by DanielTakeshi** [reply]
Thanks @mhmukadam  Regarding your Q2 answer, I would be interested in giving this a try.

As you suggest, I can start with a free-space arm planner with no collision considerations, and I will try to get the trajectory expressed in _joint space_. I patched together a minimum working example based on tutorial 04 with `GPMotionModel`. Here I am using a 6-DOF robot arm with a fixed base, which is the same one I used in my other discussion question https://github.com/facebookresearch/theseus/discussions/566 and which is [available here for download](https://drive.google.com/file/d/1R64q5MBRptyD3C-NDBUSR6g5UZVcUdH2/view?usp=sharing).

<details>
<summary>
Here is the minimum working example (138 lines long, including comments and boilerplate):
</summary>

```
import os
import random
import torch
import numpy as np
import theseus as th
from torchkin.forward_kinematics import Robot, get_forward_kinematics_fns
torch.set_printoptions(sci_mode=False, linewidth=150, precision=4)

# Bells and whistles.
dtype = torch.float32
device = "cuda:0" if torch.cuda.is_available() else "cpu"
seed = 0
torch.random.manual_seed(seed)
random.seed(seed)
np.random.seed(seed)

# -------------------------- Load the robot --------------------------- #

# See examples/inverse_kinematics.py for details.
URDF_REL_PATH = "../tests/theseus_tests/embodied/kinematics/data/ur5_no_translation_joint.urdf"
link_names = ["tool0"]
urdf_path = os.path.join(os.path.dirname(__file__), URDF_REL_PATH)
robot = Robot.from_urdf_file(urdf_path, dtype)
print(f'Just loaded robot with DOFs: {robot.dof}')

# NOTE: we are NOT using these, should we be using them?
fk, jfk_b, jfk_s = get_forward_kinematics_fns(robot, link_names)

# -------------------------- Theseus optimization --------------------------- #

# Variables for the optimization, based on 04_motion_planning.ipynb.
traj_len = 100
n_steps = traj_len - 1
total_time = 10.0
dt_val = total_time / n_steps
Qc_inv = torch.eye(robot.dof)
boundary_w = 100.0

# Create optimization variables which define a trajectory of joint angles.
poses = []
velocities = []
for i in range(traj_len):
    pose = th.Vector(robot.dof, name=f'pose_{i}', dtype=dtype)
    velocity = th.Vector(robot.dof, name=f'vel_{i}', dtype=dtype)
    poses.append(pose)
    velocities.append(velocity)

# Targets for pose boundary cost functions
start_point = th.Vector(robot.dof, name="start", dtype=dtype)
goal_point = th.Vector(robot.dof, name="goal", dtype=dtype)

# For GP dynamics cost function
dt = th.Variable(torch.tensor(dt_val).view(1, 1), name="dt")

# Cost weight to use for all GP-dynamics cost functions
gp_cost_weight = th.eb.GPCostWeight(torch.tensor(Qc_inv), dt)

# For all hard-constraints (end points pos/vel) we use a single scalar weight
# with high value
boundary_cost_weight = th.ScaleCostWeight(boundary_w)

# First, create the objective.
objective = th.Objective(dtype=dtype)

# Boundary cost functions -- initial position and velocity.
objective.add(
    th.Difference(poses[0], start_point, boundary_cost_weight, name="pose_0")
)
objective.add(
    th.Difference(
        velocities[0],
        th.Vector(tensor=torch.zeros(1, robot.dof)),
        boundary_cost_weight,
        name="vel_0",
    )
)

# Boundary cost functions -- final position and velocity.
objective.add(
    th.Difference(poses[-1], goal_point, boundary_cost_weight, name="pose_N")
)
objective.add(
    th.Difference(
        velocities[-1],
        th.Vector(tensor=torch.zeros(1, robot.dof)),
        boundary_cost_weight,
        name="vel_N",
    )
)

# GP-dynamics cost functions for smooth trajectories.
for i in range(1, traj_len):
    objective.add(
        (
            th.eb.GPMotionModel(
                poses[i - 1],
                velocities[i - 1],
                poses[i],
                velocities[i],
                dt,
                gp_cost_weight,
                name=f"gp_{i}",
            )
        )
    )

# Using the standard LM optimizer as in other Theseus examples.
optimizer = th.LevenbergMarquardt(
    objective,
    th.CholeskyDenseSolver,
    max_iterations=50,
    step_size=1.0,
)
motion_planner = th.TheseusLayer(optimizer)
motion_planner.to(device=device, dtype=dtype)

# Now run optimization with random target? Batch size 1.
start = torch.zeros((1,robot.dof), dtype=dtype).to(device)
goal = torch.rand((1,robot.dof), dtype=dtype).to(device)
print(f'\nOur starting DOFs:\n{start}')
print(f'Our target DOFs:\n{goal}\n')
planner_inputs = {
    "start": start.to(device),
    "goal": goal.to(device),
}
with torch.no_grad():
    final_values, info = motion_planner.forward(
        planner_inputs,
        optimizer_kwargs={
            "track_best_solution": True,
            "verbose": True,
            "damping": 0.1,
        }
    )

print('\nThe solution:')
for i in range(traj_len):
    print(f'i={str(i).zfill(3)}, {final_values[f"pose_{i}"]}')
```

</details>

It's like tutorial 04, except the "poses" and "velocities" are not `th.Point2` but `th.Vector` of dimension `robot.dof` (which is 6 in this example robot). Also `Qc_inv = torch.eye(robot.dof)`.

Here's the output when I run the code: https://pastebin.com/raw/uTT5dErS  I print the 6-DOF poses obtained from the optimizer which is the trajectory in joint space (well, the position component). It seems like they linearly interpolate (to some extent) between the starting all-zero DOF positions and the target, which is randomly set.

- Question 1: is this what you intended in your answer when you mentioned _"You can already easily put together a free-space arm planner using the [GPMotionModel](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/motionmodel/double_integrator.py#L179) cost if the goal is in joint space_" in your answer? Or am I missing something that you intended?
- Question 2: I am trying to connect this with the robot. I load the robot but I am only using it to get `robot.dof` to get the DOF count. I assume there must be some kind of kinematic or robot-specific constraints that we need to consider in the cost function. For example should we be using the Jacobians somehow?

Thanks again for your advice earlier.

**Reply by mhmukadam** [reply]
@DanielTakeshi that's great! 

Nice! Thanks for sharing the example. Point2 -> Vector makes sense. Feel free to start a PR if/when you want to get this merged or discuss any specifics on the code. The roughly linear interpolation makes sense as well: if you plot the joint angles (y axis) over time steps (x axis) you will see 'S-like' curve as the robot accelerates and decelerates at the start and end (assuming you set the zero velocity for the start/goal velocity cost function) and follows a linear path with constant velocity (given the GPMotionModel cost) in the middle.

Q1: Yes this is exactly what I meant.

Q2: If you want to visualize the robot you would run the joint angles through FK and get joint locations in 3D space for each joint including the end effector (this would internally be using the manipulator Jacobian). You don't need to add any additional constraints on the planner objective. If your start to goal path goes through a joint limit then in practice you would add a joint limit cost. If you want to reach a desired end effector pose (x_goal) (rather than a join angle goal) then you would add a goal cost on the final joint angle state (q_T) to match the desired pose goal after the joint angles pass through FK (c = fk(q_T) - x_goal). If you want to avoid collision you add the cost for that, and so on.

**Reply by DanielTakeshi** [reply]
Thanks @mhmukadam this is helpful!

For Q1 I was originally not sure if GPMotionModel made sense for something like Vector, hence why I was asking.

For Q2, I had assumed we would need something more detailed about the robot, but it turns out that when I did my example for Q1, to test it, I directly used those joint angle trajectories (output of Theseus) into IsaacGym where I was loading the robot. This produced smooth trajectories (with accelerating and deceleration at the beginning as you mention). Here's a video in case you're curious https://www.youtube.com/watch?v=ki6mEQM2HSc each time I change the blue sphere, that makes a new target and a new trajectory specified by Theseus (in joint angles), and the EE traverses in the curve shown in the video.

It sounds like for anything more specific, such as adding joint limit costs, I will need to make use of my own robot specific knowledge about that (which might require some trial and error).

**Reply by mhmukadam** [reply]
Thanks for sharing the robot sim video :) Great to see it seamlessly worked as expected (since all the computation happens in joint space you don't end up needing any robot specific info during optimization in this example). I assume you take the generated blue sphere's location, find the joint angle goal, and then pass it to the optimizer; or are you using the 3D sphere position as a end effector goal cost directly (as discussed in an earlier comment in this thread)?

The robot specific things like joint angle/velocity limits you could obtain from urdf and pass to the relevant [cost functions](https://github.com/gtrll/gpmp2/blob/main/gpmp2/kinematics/JointLimitFactorVector.h#L62) (once you have them setup). Shouldn't be much trail and error there (aside from param tuning).

**Reply by DanielTakeshi** [reply]
@mhmukadam Thanks for the clarifications.

In my video, I am using the joint angle goal as you suggest. So that gives me the goal joint angles (a 6-D vector) which I can use in addition to the current robot joint angles. That gives me the info I need for two parts of the cost function.

<details>
<summary>
The pseudocode would look like this:
</summary>

```
objective = th.Objective(dtype=dtype)
objective.add(
    th.Difference(poses[0], start_pose, boundary_cost_weight, name="pose_0")
)
objective.add(
    th.Difference(poses[-1], goal_pose, boundary_cost_weight, name="pose_N")
)
```

(This is not the full objective, as it's just the boundaries and would need all the intermediate poses...)
</details>

I think I should test using the 3D sphere position as an end-effector goal cost directly. That would be easier to use in practice since often we want robots to get to a 3D target position (with a 3D orientation as well). For that, do you recommend using what was in your example, i.e., this cost function was in one of the examples here https://github.com/facebookresearch/theseus/blob/main/examples/inverse_kinematics.py and just pasting the relevant portions:

```
# IK can also be solved as an optimization problem:
#      min \|log(pose^-1 * targeted_pose)\|^2
# as the following
def targeted_pose_error(optim_vars, aux_vars):
    (theta,) = optim_vars
    (targeted_pose,) = aux_vars
    pose = th.SE3(tensor=fk(theta.tensor)[0])
    return pose.local(targeted_pose)

cost_function = th.AutoDiffCostFunction(
    optim_vars=optim_vars,
    err_fn=targeted_pose_error,
    dim=6,
    aux_vars=aux_vars,
    name="targeted_pose_error",
    cost_weight=th.ScaleCostWeight(torch.ones([1], dtype=dtype)),
    autograd_mode="vmap",
)
```

Additionally, one thing that came up when I was doing these tests is: I would like the robot to be fixed but at a different starting pose. For example the robot might be above the surface and pointing in a specific direction.

From the IK examples the `fk` function is based on the `robot` object:

```
URDF_REL_PATH = "../tests/theseus_tests/embodied/kinematics/data/panda_no_gripper.urdf"
urdf_path = os.path.join(os.path.dirname(__file__), URDF_REL_PATH)
robot = Robot.from_urdf_file(urdf_path, dtype)
link_names = ["panda_virtual_ee_link"]
fk, jfk_b, jfk_s = get_forward_kinematics_fns(robot, link_names)
```

And that class will load the URDF file.
https://github.com/facebookresearch/theseus/blob/main/torchkin/torchkin/robot.py
But is there a way to just make the starting pose of the robot different (some kind of API method)? Then ideally everything just propagates from there. Otherwise the fk will assume all the poses based on the URDF. (I could modify the URDF, but many times we want to test robots in different initial configurations.) 

Thanks again for your advice and assistance.

**Reply by DanielTakeshi** [reply]
After some testing, this is my effort to do the following (basically what I described in my last post) but technically using the full pose instead of just the position:

> using the 3D sphere position as a end effector goal cost directly (as discussed in an earlier comment in this thread)?

<details>
<summary>
So now, the relevant code for the cost function portion looks like this (only showing the cost function, code is not runnable but hopefully this makes it clear what is happening, a lot is straight from Theseus tutorials):
</summary>

```
traj_len = 100
n_steps = traj_len - 1
total_time = 10.0
dt_val = total_time / n_steps
Qc_inv = torch.eye(robot.dof)
boundary_w = 100.0

# Create optimization variables which define a trajectory of joint angles.
poses = []
velocities = []
for i in range(traj_len):
    pose = th.Vector(robot.dof, name=f'pose_{i}', dtype=dtype)
    velocity = th.Vector(robot.dof, name=f'vel_{i}', dtype=dtype)
    poses.append(pose)
    velocities.append(velocity)

# Starting joint angle targets for the initial pose boundary cost.
start_joints = th.Vector(robot.dof, name="start_j", dtype=dtype)

# For GP dynamics cost function
dt = th.Variable(torch.tensor(dt_val).view(1, 1), name="dt")

# Cost weight to use for all GP-dynamics cost functions
gp_cost_weight = th.eb.GPCostWeight(torch.tensor(Qc_inv), dt)

# For all hard-constraints (end points pos/vel) use scalar weight w/high value
boundary_cost_weight = th.ScaleCostWeight(boundary_w)

# Create the objective. Add the initial pose, we already know what that is.
objective = th.Objective(dtype=dtype)
objective.add(
    th.Difference(poses[0], start_joints, boundary_cost_weight, name="pose_0")
)
objective.add(
    th.Difference(
        velocities[0],
        th.Vector(tensor=torch.zeros(1, robot.dof)),
        boundary_cost_weight,
        name="vel_0",
    )
)
objective.add(
    th.Difference(
        velocities[-1],
        th.Vector(tensor=torch.zeros(1, robot.dof)),
        boundary_cost_weight,
        name="vel_N",
    )
)

# GP-dynamics cost functions for smooth trajectories.
for i in range(1, traj_len):
    objective.add(
        (
            th.eb.GPMotionModel(
                poses[i - 1],
                velocities[i - 1],
                poses[i],
                velocities[i],
                dt,
                gp_cost_weight,
                name=f"gp_{i}",
            )
        )
    )

# The target is not specified in joints! Don't use this.
# objective.add(
#     th.Difference(poses[-1], goal_pose, boundary_cost_weight, name="pose_N")
# )
#
# Instead frame it as the following, where we use target pose error.
# IK can be solved as an optimization problem:
#      min \|log(pose^-1 * targeted_pose)\|^2
# as the following
def targeted_pose_error(optim_vars, aux_vars):
    (theta,) = optim_vars
    (targeted_pose,) = aux_vars
    pose = th.SE3(tensor=fk(theta.tensor)[0])
    return pose.local(targeted_pose)

# Stuff that we need for targeted_pose_error. Auxiliary variable is the target EE
# pose which we can treat as ground-truth and we want to find optim_var_theta.
optim_vars = (poses[-1],)  # I think this is it?
goal_ee_pose = torch.zeros((1,3,4))  # Assign to this via 'goal_ee' key.
aux_vars = (th.SE3(tensor=goal_ee_pose, name="goal_ee"),)

# Finally, add the cost function for the FINAL pose error. See Theseus IK example.
cost_function = th.AutoDiffCostFunction(
    optim_vars,
    targeted_pose_error,
    dim=6,
    aux_vars=aux_vars,
    name="targeted_pose_error",
    cost_weight=th.ScaleCostWeight(1 * torch.ones([1], dtype=dtype)),
    autograd_mode="vmap",
)
objective.add(cost_function)
```

</details>

My videos/tests are the result of the procedure:

- I sample different joint angles of the robot, and run forward kinematics to get the EE position and orientation. Those become the future EE target. By doing so, we know that the EE targets are feasible.
- The cost function (see code segment above) consists of: (1) velocities should be 0 at the boundaries, (2) the first joint angle should match the starting joint angle, which we can get from any robot by querying its state, (3) we still use the Gaussian Process cost for smooth trajectories, (4) but different this time, we use `targeted_pose_error` as the cost function so that the last joint angle is such that it should lead to the EE pose matching the target pose.
- I sequentially iterate through the different targets, and for each target, run Theseus optimization to get a list of joint angles for the robot at each time step (uses the cost function). Then tell the robot to follow the trajectory, going through each waypoint.

(This is the same as my last video I shared, the only major change is that the cost function is different for the **last** joint angle.)


One thing I noticed: the cost function for the ending pose needs to be tuned different. With the boundary cost function of 100, I get very wacky results: https://youtu.be/f4Zi2F9E6oM

I had to tune it down to 10:

```
cost_function = th.AutoDiffCostFunction(
    optim_vars,
    targeted_pose_error,
    dim=6,
    aux_vars=aux_vars,
    name="targeted_pose_error",
    cost_weight=th.ScaleCostWeight(10 * torch.ones([1], dtype=dtype)),
    autograd_mode="vmap",
)
```

Then the results seemed OK. https://youtu.be/s4Rm9IQd3NU
I assume this is because the pose difference isn't really on the same scale as a vector difference, potentially due to what it means to "take a difference between two rotations".  So it has to be tuned different compared to the boundary cost values for the other parts of the cost. Does that seem reasonable?

**Reply by mhmukadam** [reply]
That looks great! Your cost functions are correct and it's reasonable that some weight tuning is needed based on the composition of costs in the objective.

> I would like the robot to be fixed but at a different starting pose. For example the robot might be above the surface and pointing in a specific direction.
> 
> From the IK examples the fk function is based on the robot object:
> 
> ```
> URDF_REL_PATH = "../tests/theseus_tests/embodied/kinematics/data/panda_no_gripper.urdf"
> urdf_path = os.path.join(os.path.dirname(__file__), URDF_REL_PATH)
> robot = Robot.from_urdf_file(urdf_path, dtype)
> link_names = ["panda_virtual_ee_link"]
> fk, jfk_b, jfk_s = get_forward_kinematics_fns(robot, link_names)
> ```
> And that class will load the URDF file.
> https://github.com/facebookresearch/theseus/blob/main/torchkin/torchkin/robot.py
> But is there a way to just make the starting pose of the robot different (some kind of API method)? Then ideally everything just propagates from there. Otherwise the fk will assume all the poses based on the URDF. (I could modify the URDF, but many times we want to test robots in different initial configurations.)

For your question about starting the robot at a different base pose than the origin: @fantaosha can you share a pointer? Hopefully it's easy to add this feature if it's currently missing.

**Reply by DanielTakeshi** [reply]
Thanks again @mhmukadam  I also have a cleaner video here https://youtu.be/b-fArLteqZQ which now shows the orientation. Basically a "good" IK will get the last link so that it is "facing" from the green to blue spheres, so visually, an arrow starting from the green sphere and going to the blue sphere points in the desired direction of rotation. The video shows really good reaching results. This also has extra joint limit cost functions as you suggest, and I'm using a per-time step hinge cost on the joint angle positions (basically making them obey joint limits) https://github.com/facebookresearch/theseus/blob/dc710c25880f0c7b64709b72136308e28da46566/theseus/embodied/motionmodel/misc.py#L14 

Yes, the pointer about starting the robot at a different base pose would be helpful if possible, thanks.

**Reply by mhmukadam** [reply]
Nice great to see the polished result now with joint limit costs. Would you like to make a PR and include this in the repo as an example?

**Reply by fantaosha** [reply]
If I understand correctly, you would like a custom root pose other than identity. Is this correct? Thanks!

**Reply by DanielTakeshi** [reply]
Thanks @mhmukadam , there is one thing I probably want to do before I submit a PR: figure out a good "initial" trajectory. I think in my code the joint angles start at all 0 (though I have to double check) and it might be helpful to try a better initialization scheme.

@fantaosha Basically what I'd like to do is something like I show in this video https://www.youtube.com/watch?v=b-fArLteqZQ  but where the robot base (or root) starts at a different pose.  In the below code from IK, we are loading `robot` from the URDF and then defining a forward kinematics function from that.

https://github.com/facebookresearch/theseus/blob/dd4356b7688c2610be9f99f7f67ec532a6677fcd/examples/inverse_kinematics.py#L28-L43

Is there a way to do something like: `robot.adjust_base_pose( translation, orientation )` so that we can adjust the robot (base, or "root") pose before calling `get_forward_kinematics_fns`, and then have the forward kinematics function accurately reflect that? For example if we adjusted the base pose so that the z coordinate of the robot base starts at 0.1 meters above ground, then all the forward kinematics functions for EE poses should result in z coordinate values of +0.1 compared to the default robot pose.

I'm not sure how easy/hard it would be to integrate in the current code, so just inquiring about this. Thanks!

**Reply by mhmukadam** [reply]
Sounds good on the PR. For traj initialization there are a few things we have tried in the past (each has it's pros/cons): (1) straight line in joint angle space between start and goal (possible if you have a joint goal), (2) all traj is at the start joint angles, (3) random/zero init.

**Reply by DanielTakeshi** [reply]
@fantaosha Just revisiting this discussion, do you know if there is function to support "a custom root pose other than identity"?

**Reply by fantaosha** [reply]
Hi @DanielTakeshi, sorry for the late replay. Unfortunately, there is no such function to set the root pose. But I think it is quite straightforward to write such a function by yourselves using group composition and adjoint transformation. Thanks!
